src/thunderbots/software/ai/primitive/directwheels_primitive.cpp

```diff
@@ -0,0 +1,120 @@
+#include "ai/primitive/directwheels_primitive.h"
+
+#include "util/logger/init.h"
+
+const std::string DirectWheelsPrimitive::PRIMITIVE_NAME = "DirectWheels Primitive";
+
+DirectWheelsPrimitive::DirectWheelsPrimitive(
+    unsigned int robot_id, int front_left_wheel_power, int back_left_wheel_power,
+    int front_right_wheel_power, int back_right_wheel_power, double dribbler_rpm)
+    : robot_id(robot_id),
+      front_left_wheel_power(front_left_wheel_power),
+      back_left_wheel_power(back_left_wheel_power),
+      front_right_wheel_power(front_right_wheel_power),
+      back_right_wheel_power(back_right_wheel_power),
+      dribbler_rpm(dribbler_rpm)
+{
+    // Clamp wheel power if out of range, log a warning
+    if (abs(front_left_wheel_power) > 255)
+    {
+        this->front_left_wheel_power = std::clamp(front_left_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: front left wheel power out of bounds";
+    }
+    if (abs(back_left_wheel_power) > 255)
+    {
+        this->back_left_wheel_power = std::clamp(back_left_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: back left wheel power out of bounds";
+    }
+    if (abs(front_right_wheel_power) > 255)
+    {
+        this->front_right_wheel_power = std::clamp(front_right_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: front right wheel power out of bounds";
+    }
+    if (abs(back_right_wheel_power) > 255)
+    {
+        this->back_right_wheel_power = std::clamp(back_right_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: back right wheel power out of bounds";
+    }
+}
+
+DirectWheelsPrimitive::DirectWheelsPrimitive(
+    const thunderbots_msgs::Primitive &primitive_msg)
+{
+    validatePrimitiveMessage(primitive_msg, getPrimitiveName());
+
+    robot_id                = primitive_msg.robot_id;
+    front_left_wheel_power  = int(primitive_msg.parameters.at(0));
+    back_left_wheel_power   = int(primitive_msg.parameters.at(1));
+    front_right_wheel_power = int(primitive_msg.parameters.at(2));
+    back_right_wheel_power  = int(primitive_msg.parameters.at(3));
+
```

Comment:
The `dribbler_rpm` is not being assigned here. @SupoWupo this is the cause of those test failures.